binderhub/builder.py

```diff
@@ -152,55 +152,57 @@ def get(self, provider_prefix, spec):
             ref=ref,
             image_name=image_name,
             push_secret=push_secret,
-            builder_image=self.settings['builder_image_spec']
+            builder_image=self.settings['builder_image_spec'],
         )
 
-        # FIXME: use a thread pool
-        build_thread = threading.Thread(target=build.submit)
-        log_thread = threading.Thread(target=build.stream_logs)
+        pool = self.settings['build_pool']
+        pool.submit(build.submit)
+        log_future = None
 
-        build_thread.start()
+        # yield initial waiting event
+        try:
+            yield self.emit({
+                'phase': 'waiting',
+                'message': 'Waiting for build to start...\n',
+            })
+        except StreamClosedError:
+            # Client has gone away!
+            return
 
         done = False
-
-        while True:
-            # FIXME: use Futures
-            try:
-                progress = q.get_nowait()
-            except Empty:
-                yield gen.sleep(0.5)
-                continue
+        while not done:
+            progress = yield q.get()
```

Comment:
Does this work? I don't see anything about generator / async support in https://docs.python.org/3/library/queue.html

Reply:
This is not a stdlib `queue.Queue`, it's a tornado `queues.Queue`, which is like a Queue except it's async (hence `yield`).